weather/src/test/java/com/oddle/app/weather/controller/WeatherControllerTest.java

```diff
@@ -0,0 +1,202 @@
+package com.oddle.app.weather.controller;
+
+import static org.hamcrest.CoreMatchers.any;
+import static org.hamcrest.CoreMatchers.is;
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+import com.google.gson.Gson;
+import com.oddle.app.weather.dto.request.weather.save.EditWeatherDataRequest;
+import com.oddle.app.weather.dto.request.weather.save.SaveWeatherDataRequest;
+import com.oddle.app.weather.dto.response.HistoryWeatherResponse;
+import com.oddle.app.weather.dto.response.WeatherResponse;
+import com.oddle.app.weather.dto.response.openweather.OWCurrentWeatherResponse;
+import com.oddle.app.weather.model.weatherdata.Weather;
+import com.oddle.app.weather.service.CityWeatherService;
+import com.oddle.app.weather.service.WeatherService;
+
+
+
+
+@SpringBootTest
+@AutoConfigureMockMvc
+public class WeatherControllerTest {
+
+	
+    @Autowired
+    MockMvc mockMvc;
+    
+    @MockBean
+    @Qualifier("OpenWeatherService")
+    CityWeatherService cityWeatherService;
+    
+    @MockBean
+    @Qualifier("OpenWeatherService")
+    WeatherService weatherService;
+    
+    @Test
+    public void testOpenWeatherCurrentApi() throws Exception {
+    	String name = "manila";
+    	OWCurrentWeatherResponse response = new OWCurrentWeatherResponse();
+    	response.setName(name);
+    	
+		
+		when(cityWeatherService.getCurrentWeather(name)).thenReturn(response);
```

Comment:
in this case, we are not actually testing integration with OpenWeather right? Since the result is mocked and always correct

Reply:
yes did'nt include the api layer for testing, i only include service and controller layer test for future proofing of the flow from service to controller when maintaining and enhancing is sufficient,
for testing integration we are testing it usually on postman and swagger when the service and controller layer is completed. the cause of that practice is sometimes the sandbox testing api of the 3rd party sometimes is down or number of api calls in testings environment sometimes has limits. affecting the build of our cicd in testing env., that's why

I can add integration unit test if you want

Reply:
@jay0911 Thanks for your clarification. No, it's not required to add the integration, we got what you mean for mocking the API call and it makes sense.